src/xrpld/app/tx/detail/OnChainCredentials.cpp

```diff
@@ -0,0 +1,454 @@
+    sleCred->setAccountID(sfSubject, subj);
+    sleCred->setAccountID(sfIssuer, issuer);
+    sleCred->setFieldVL(sfCredentialType, credType);
+
+    if (ctx_.tx.isFieldPresent(sfSignature))
+        sleCred->setFieldVL(sfSignature, ctx_.tx.getFieldVL(sfSignature));
+
+    auto const optExp = ctx_.tx[~sfExpiration];
+    if (optExp)
+    {
+        if (closeTime > *optExp)
+        {
+            JLOG(j_.trace()) << "Malformed transaction: "
+                                "Expiration time is in the past.";
+            return tecEXPIRED;
+        }
+
+        sleCred->setFieldU32(sfExpiration, ctx_.tx.getFieldU32(sfExpiration));
+    }
+
+    if (ctx_.tx.isFieldPresent(sfURI))
+        sleCred->setFieldVL(sfURI, ctx_.tx.getFieldVL(sfURI));
+
+    if (issPresent)
+        sleCred->setFieldU32(sfFlags, lsfAccepted);
+
+    view().insert(sleCred);
+    auto const page = view().dirInsert(
+        keylet::ownerDir(account_), kCred, describeOwnerDir(account_));
+
+    JLOG(j_.trace()) << "Adding Credential to owner directory "
+                     << to_string(kCred.key) << ": "
+                     << (page ? "success" : "failure");
+
+    if (!page)
+        return tecDIR_FULL;
+
+    sleCred->setFieldU64(sfOwnerNode, *page);
+
+    adjustOwnerCount(view(), sleOwner, 1, j_);
+    view().update(sleOwner);
+
+    return tesSUCCESS;
+}
+
+// ------- DELETE --------------------------
+NotTEC
+CredentialDelete::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureOnChainCredentials))
+        return temDISABLED;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    auto& tx = ctx.tx;
+    auto& j = ctx.j;
+
+    auto const optSubj = tx[~sfSubject];
+    auto const optIss = tx[~sfIssuer];
+
+    if (!optSubj && !optIss)
+    {
+        // Neither field is present, the transaction is malformed.
+        JLOG(j.trace()) << "Malformed transaction: "
+                           "Invalid Subject and Issuer fields combination.";
+        return temMALFORMED;
+    }
+
+    // Make sure that the passed account is valid.
+    if ((optSubj && !AccountID(*optSubj)) || (optIss && !AccountID(*optIss)))
+    {
+        JLOG(j.trace()) << "Malformed transaction: Subject or Issuer "
+                           "field zeroed.";
+        return temINVALID_ACCOUNT_ID;
+    }
+
+    return preflight2(ctx);
+}
+
+TER
+CredentialDelete::preclaim(PreclaimContext const& ctx)
+{
+    AccountID const acc{ctx.tx[sfAccount]};
+
+    auto const& txCredType(ctx.tx[sfCredentialType]);
+    Blob const credType{txCredType.begin(), txCredType.end()};
+
+    auto const subj = ctx.tx[~sfSubject];
+    auto const iss = ctx.tx[~sfIssuer];
+    auto const kCred =
+        keylet::credential(subj ? *subj : acc, iss ? *iss : acc, credType);
+    auto sleCred = ctx.view.read(kCred);
+
+    if (!sleCred)
+        return tecNO_ENTRY;
+
+    return tesSUCCESS;
+}
+
+TER
+CredentialDelete::deleteSLE(
```

Comment:
If this is the same helper function from `DID.cpp`, it might be worth making it a helper function in `View.cpp`.

Reply:
It used 3-4 times in the code, it doesn't require such wide namespace.